### DISCUSSION TITLE
custom on_attach for every server

### DISCUSSION THREAD

**Discussion by bugabinga** [VonHeikemen/lsp-zero.nvim#24] [Q&A] [answered]
How would you go about extending the `on_attach` of every lsp server?
Pseudo example:

```lua
local lsp_opts = {
  flags = {
    debounce_text_changes = 150,
  }
}

lsp.setup_servers({
  '*', -- ANY SERVER
  opts = lsp_opts
})
```

I was trying to integrate [nvim-navic](https://github.com/SmiteshP/nvim-navic) with lsp-zero.
nvim-navic wants to be attached to exaclty one lsp per buffer, so the above approach would probably not even work (because I added null-lsp, which is often attached as well), but I was curious none the less.

lsp-zero is lovely so far, thanks for sharing it !

**Comment by VonHeikemen** [comment, marked as answer]
You can use the `.on_attach` function. 

To use navic, you can try to create a buffer local variable and use that to control the function.

```lua
local lsp = require('lsp-zero')
lsp.preset('recommended')

lsp.on_attach(function(client, bufnr)
  if vim.b.lsp_attached then return  end
  vim.b.lsp_attached = true

  require('nvim-navic').attach(client, bufnr)
end)

lsp.setup()
```

**Reply by bugabinga** [reply]
ah my bad. it is right here in the docs. i just missed it. thanks a lot :)
